src/transport/http2.rs

```diff
@@ -166,29 +161,50 @@ impl Default for ChannelConfig {
 #[derive(Debug)]
 pub struct ServerChannel<In: RpcMessage, Out: RpcMessage> {
     /// The channel.
-    channel: Receiver<(Receiver<In>, Sender<Out>)>,
+    channel: Receiver<InternalChannel>,
+    /// The configuration.
+    config: Arc<ChannelConfig>,
     /// The sender to stop the server.
     ///
     /// We never send anything over this really, simply dropping it makes the receiver
     /// complete and will shut down the hyper server.
     stop_tx: mpsc::Sender<()>,
+    /// Phantom data for in and out
+    _p: PhantomData<(In, Out)>,
 }
 
 impl<In: RpcMessage, Out: RpcMessage> ServerChannel<In, Out> {
     /// Creates a server listening on the [`SocketAddr`], with the default configuration.
     pub fn serve(addr: &SocketAddr) -> hyper::Result<Self> {
-        Self::serve_with_config(addr, ChannelConfig::default())
+        let mut addr_incoming = AddrIncoming::bind(addr)?;
+        addr_incoming.set_nodelay(true);
+        Self::serve0(addr_incoming, Default::default())
     }
 
-    /// Creates a server listening on the [`SocketAddr`] with a custom configuration.
-    pub fn serve_with_config(addr: &SocketAddr, config: ChannelConfig) -> hyper::Result<Self> {
+    /// Creates a server listener given an arbitrary incoming and a custom configuration.
+    pub fn serve_with_incoming<I>(incoming: I, config: Arc<ChannelConfig>) -> hyper::Result<Self>
+    where
+        I: hyper::server::accept::Accept + Send + 'static,
+        I::Conn: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+        I::Error: Into<Box<dyn error::Error + Send + Sync>>,
+    {
+        Self::serve0(incoming, config)
+    }
+
+    fn serve0<I>(incoming: I, config: Arc<ChannelConfig>) -> hyper::Result<Self>
+    where
+        I: hyper::server::accept::Accept + Send + 'static,
+        I::Conn: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+        I::Error: Into<Box<dyn error::Error + Send + Sync>>,
+    {
         let (accept_tx, accept_rx) = flume::bounded(32);
 
         // The hyper "MakeService" which is called for each connection that is made to the
         // server.  It creates another Service which handles a single request.
-        let service = make_service_fn(move |socket: &AddrStream| {
-            let remote_addr = socket.remote_addr();
-            event!(Level::TRACE, "Connection from {:?}", remote_addr);
+        let service = make_service_fn(move |_socket: &I::Conn| {
+            // TODO: log remote address
+            // let remote_addr = socket.remote_addr();
+            // event!(Level::TRACE, "Connection from {:?}", remote_addr);
```

Comment:
In the generic version, I::Conn does no longer have a remote addr.

Reply:
I put in a fix, but since I don't like the fix enough to make it public, I made the entire custom method private again.

We could also completely remove it, but that would complicate things with subsequent PRs.

Reply:
i think this way around is fine